lib/ex_aws/cloudfront/custom_policy.ex

```diff
@@ -0,0 +1,125 @@
+  @enforce_keys [:url, :date_less_than]
+
+  defstruct [:url, :date_less_than, :date_greater_than, :ip_address]
+
+  @doc """
+  Create a Custom Policy.
+  """
+  def create(url), do: create(url, ExAws.Utils.now_in_seconds + 1800)
+  def create(url, %DateTime{} = date_less_than), do: create(url, DateTime.to_unix(date_less_than))
+  def create(url, date_less_than) when is_binary(url) and is_integer(date_less_than) do
+    %__MODULE__{
+      url: url,
+      date_less_than: date_less_than
+    }
+  end
+
+  @doc """
+  Puts a beginning date and time in Unix time format and UTC.
+  """
+  def put_date_greater_than(%__MODULE__{} = policy, %DateTime{} = value) do
+    put_date_greater_than(policy, DateTime.to_unix(value))
+  end
+
+  def put_date_greater_than(%__MODULE__{} = policy, value) when is_nil(value) or is_integer(value) do
+    %__MODULE__{ policy | date_greater_than: value }
+  end
+
+  @doc """
+  Puts an IP address.
+  """
+  def put_ip_address(%__MODULE__{} = policy, value) when is_nil(value) or is_binary(value) do
+    %__MODULE__{ policy | ip_address: value }
+  end
+end
+
+defimpl ExAws.CloudFront.Policy, for: ExAws.CloudFront.CustomPolicy do
+  import ExAws.CloudFront.Utils
+
+  @doc """
+  Create a Signed URL Using a Custom Policy.
+  """
+  def get_signed_url(custom_policy, keypair_id, private_key) do
+    get_signed_url(custom_policy, fn payload ->
+      [
+        {"Policy", payload |> aws_encode64},
+        {"Signature", payload |> create_signature(private_key) |> aws_encode64},
+        {"Key-Pair-Id", keypair_id}
+      ]
+    end)
+  end
+
+  @doc """
+  Creating a Signature for a Signed Cookie That Uses a Custom Policy.
+  """
+  def get_signed_cookies(custom_policy, keypair_id, private_key) do
+    get_signed_cookies(custom_policy, fn payload ->
+      %{
+        "CloudFront-Policy" => payload |> aws_encode64,
+        "CloudFront-Signature" => payload |> create_signature(private_key) |> aws_encode64,
+        "CloudFront-Key-Pair-Id" => keypair_id
+      }
+    end)
+  end
+
+  @doc """
+  Create a Policy Statement for a Signed URL That Uses a Custom Policy.
+  """
+  def to_statement(%{
+    url: url,
+    date_less_than: date_less_than,
+    date_greater_than: date_greater_than,
+    ip_address: ip_address
+  }) do
+    unless date_less_than < 2147483647 do
+      raise ArgumentError, message:
+        "`date_less_than` must be less than 2147483647 (January 19, 2038 03:14:08 GMT)"
+    end
+
+    unless date_less_than > ExAws.Utils.now_in_seconds do
+      raise ArgumentError, message: "`date_less_than` must be after the current time"
+    end
+
+    unless is_nil(date_greater_than) do
+      unless date_greater_than < date_less_than do
+        raise ArgumentError, message: "`date_greater_than` must be before the `date_less_than`"
+      end
+
+      unless date_greater_than > ExAws.Utils.now_in_seconds do
+        raise ArgumentError, message: "`date_greater_than` must be after the current time"
+      end
+    end
```

Comment:
I'm unfortunately not the biggest fan of the idea that the default error handling here is via exceptions. At a minimum we ought to help people expect that by using `!` in the function name, but either way I think it'd be better of the function signature was `@spec to_statement(policy) :: {:ok, statement} | {:error, reason}`